luarules/gadgets/unit_shield_behaviour.lua

```diff
@@ -0,0 +1,334 @@
+function gadget:GetInfo()
+	return {
+		name = "Shield Behaviour",
+		desc = "Overrides default shield engine behavior. Defines downtime.",
+		author = "SethDGamre",
+		layer = 1,
+		enabled = true
+	}
+end
+
+if not Spring.GetModOptions().shieldsrework then return false end
+if not gadgetHandler:IsSyncedCode() then return end
+
+---- Optional unit customParams ----
+-- shield_downtime = <number in seconds>, if not set defaults to defaultDowntime
+-- shield_aoe_penetration = bool, if true then AOE damage will hurt units within the shield radius
+
+local defaultDowntime = 1
+
+-- To save on performance, do not perform AoE damage mitigation checks below this threshold, value chosen empirically to negate laser AoE from a Sumo
+local aoeIgnoreThreshold = 11
+
+-- Units half-in/half-out of a shield should not be protected, so need a buffer of non-coverage near the edge, value chosen empirically through testing to avoid having to look up collision volumes
+local radiusExclusionBuffer = 10
+
+local spGetUnitShieldState = Spring.GetUnitShieldState
+local spSetUnitShieldState = Spring.SetUnitShieldState
+local spGetGameSeconds = Spring.GetGameSeconds
+local spSetUnitShieldRechargeDelay = Spring.SetUnitShieldRechargeDelay
+local spDeleteProjectile = Spring.DeleteProjectile
+local spGetProjectileDefID = Spring.GetProjectileDefID
+local spGetUnitPosition = Spring.GetUnitPosition
+local spGetUnitsInSphere = Spring.GetUnitsInSphere
+local spGetProjectilesInRectangle = Spring.GetProjectilesInRectangle
+local spGetUnitIsStunned = Spring.GetUnitIsStunned
+local spAreTeamsAllied = Spring.AreTeamsAllied
```

Comment:
What you changed here...

I knew this command existed I intentionally compiled a list of all the allied teams at runtime to prevent having to use a call in every single time an allied check was performed

Reply:
I didn't see any noticeable performance differences, and it is better to use the built-in engine functions than to re-implement them locally.

Reply:
Okay! Seems legit